### PR TITLE
feat: Create job and Watch job status for minioJob

### PR DESCRIPTION
**Pull request by jiuker**
watch the minioJob
1.handle the minioJob
2.check ref tenant
3.check sa
4.create the job
5.generate the status to minioJob according to the intervalJob status

watch the job
1.update the intervalJob status

How to test:
1. deploy operator with env`OPERATOR_STS_ENABLED = on`
2. deploy tenant
3. deploy minioJob
```yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  name: mc-job-sa
---
apiVersion: sts.min.io/v1alpha1
kind: PolicyBinding
metadata:
  name: mc-job-bingding
spec:
  application:
    serviceaccount: mc-job-sa
  policies:
    - consoleAdmin
---
apiVersion: job.min.io/v1alpha1
kind: MinIOJob
metadata:
  name: minio-test-job
spec:
  serviceAccountName: mc-job-sa
  tenant:
    name: mytest-minio
  commands:
    - op: make-bucket
      args:
        name: memes
    - name: add-my-user-1
      op: admin/user/add
      args:
        user: daniel
        password: daniel123
    - name: add-my-user-2
      op: admin/user/add
      args:
        user: pedro
        password: daniel123
    - name: add-my-policy
      op: admin/policy/create
      args:
        name: memes-access
        policy: |
          {
              "Version": "2012-10-17",
              "Statement": [
                  {
                      "Effect": "Allow",
                      "Action": [
                          "s3:*"
                      ],
                      "Resource": [
                          "arn:aws:s3:::memes",
                          "arn:aws:s3:::memes/*"
                      ]
                  }
              ]
          }
    - op: admin/policy/attach
      dependsOn:
        - add-my-user-1
        - add-my-user-2
        - add-my-policy
      args:
        policy: memes-access
        user: daniel
    - op: admin/policy/attach
      dependsOn:
        - add-my-user-1
        - add-my-user-2
        - add-my-policy
      args:
        policy: memes-access
        user: pedro
```

Failed example:
```yaml
apiVersion: job.min.io/v1alpha1
kind: MinIOJob
metadata:
 ...
status:
  commands:
    - message: ''
      name: command-0
      result: success
    - message: ''
      name: add-my-user-1
      result: success
    - message: ''
      name: add-my-user-2
      result: success
    - message: ''
      name: add-my-policy
      result: success
    - message: Job has reached the specified backoff limit
      name: command-4
      result: failed
    - message: Job has reached the specified backoff limit
      name: command-5
      result: failed
  message: Job has reached the specified backoff limit
  phase: Failed
spec:
 ...
```
Running example:
```yaml
apiVersion: job.min.io/v1alpha1
kind: MinIOJob
metadata:
 ...
status:
  commands:
    - message: ''
      name: command-0
      result: success
    - message: ''
      name: add-my-user-1
      result: success
    - message: ''
      name: add-my-user-2
      result: success
    - message: ''
      name: add-my-policy
      result: success
    - message: ''
      name: command-4
      result: running
    - message: Job has reached the specified backoff limit
      name: command-5
      result: failed
  message: ''
  phase: Running
spec:
 ...
```

### CODE REVIEW

**Comment by cniackz** [comment]
Yes, I like this approach. Operation Alias to MC Command function that can translate from:

```yaml
    - op: make-bucket
      args:
        name: memes
```

to:

```shell
mc mb ALIAS/memes
```

**Comment by cniackz** [comment]
Please continue with these changes; they look good to me. Just ensure that you address the lint issues and mark the pull request as ready for testing. After that, we can conduct another round of review.

Thank you very much @jiuker 🥳

**Comment by harshavardhana** [comment]
This looks like a useful change // cc @pjuarezd

File: pkg/controller/job-controller.go [hunk 1]
```diff
@@ -107,12 +157,33 @@ func NewJobController(
...
-			// todo record the job status.
+			val, ok := globalIntervalJobStatus.Load(fmt.Sprintf("%s/%s", newJob.GetNamespace(), jobCRName))
+			if ok {
+				intervalJob := val.(*MinIOIntervalJob)
+				command, ok := intervalJob.CommandMap[jobName]
+				if ok {
+					if newJob.Status.Succeeded > 0 {
+						command.setStatus(true, "")
+					} else {
+						for _, condtion := range newJob.Status.Conditions {
```
Comment (shtripat) [inline pkg/controller/job-controller.go:177]:
```suggestion
						for _, condition := range newJob.Status.Conditions {
```

File: pkg/controller/job-controller.go [hunk 2]
```diff
@@ -163,11 +234,22 @@ func (c *JobController) SyncHandler(key string) (Result, error) {
 	err := c.k8sClient.Get(ctx, client.ObjectKeyFromObject(&jobCR), &jobCR)
 	if err != nil {
 		// job cr have gone
+		globalIntervalJobStatus.Delete(fmt.Sprintf("%s/%s", jobCR.Namespace, jobCR.Name))
 		if errors.IsNotFound(err) {
 			return WrapResult(Result{}, nil)
 		}
 		return WrapResult(Result{}, err)
 	}
+	// if job is success, do nothing
```
Comment (shtripat) [inline pkg/controller/job-controller.go:243]: we are actually deleting ;)

File: pkg/controller/job-controller.go [hunk 3]
```diff
@@ -203,16 +285,428 @@ func (c *JobController) SyncHandler(key string) (Result, error) {
...
+	err = intervalJob.createCommandJob(ctx, c.k8sClient)
+	if err != nil {
+		jobCR.Status.Phase = MinioJobPhaseError
+		jobCR.Status.Message = fmt.Sprintf("Create job error:%v", err)
+		err = c.updateJobStatus(ctx, &jobCR)
+		return WrapResult(Result{}, err)
 	}
+	// update status
+	status := intervalJob.getMinioJobStatus(ctx)
+	jobCR.Status = status
```
Comment (shtripat) [inline pkg/controller/job-controller.go:297]:
```suggestion
	jobCR.Status = intervalJob.getMinioJobStatus(ctx)
```

File: pkg/controller/job-controller.go [hunk 4]
```diff
@@ -203,16 +285,428 @@ func (c *JobController) SyncHandler(key string) (Result, error) {
...
+	Content string
+}
+
+// MinIOIntervalJobCommand - Job run command
+type MinIOIntervalJobCommand struct {
+	Locker      sync.RWMutex
+	JobName     string
+	MCOperation string
+	Command     string
+	DepnedOn    []string
```
Comment (shtripat) [inline pkg/controller/job-controller.go:339]:
```suggestion
	DependsOn    []string
```

File: pkg/controller/job-controller.go [hunk 5]
```diff
@@ -203,16 +285,428 @@ func (c *JobController) SyncHandler(key string) (Result, error) {
...
+type MinIOIntervalJobCommandFile struct {
+	Name    string
+	Ext     string
+	Dir     string
+	Content string
+}
+
+// MinIOIntervalJobCommand - Job run command
+type MinIOIntervalJobCommand struct {
+	Locker      sync.RWMutex
```
Comment (shtripat) [inline pkg/controller/job-controller.go:335]:
```suggestion
	mutex      sync.RWMutex
```
a private one

File: pkg/controller/job-controller.go [hunk 6]
```diff
@@ -203,16 +285,428 @@ func (c *JobController) SyncHandler(key string) (Result, error) {
...
+				},
+			},
+		},
+	}
+	if jobCR.Spec.FailureStrategy == v1alpha1.StopOnFailure {
+		job.Spec.Template.Spec.RestartPolicy = corev1.RestartPolicyNever
+	} else {
+		job.Spec.Template.Spec.RestartPolicy = corev1.RestartPolicyOnFailure
+	}
+	if len(jobCommand.Files) != 0 {
```
Comment (shtripat) [inline pkg/controller/job-controller.go:443]:
```suggestion
	if len(jobCommand.Files) > 0 {
```

File: pkg/controller/job-controller.go [hunk 7]
```diff
@@ -203,16 +285,428 @@ func (c *JobController) SyncHandler(key string) (Result, error) {
...
+	jobCommand.Created = true
+	jobCommand.Locker.Unlock()
+	return nil
+}
+
+// MinIOIntervalJob - Interval Job
+type MinIOIntervalJob struct {
+	// to see if that change
+	JobCR      *v1alpha1.MinIOJob
+	Locker     sync.Mutex
```
Comment (shtripat) [inline pkg/controller/job-controller.go:494]:
```suggestion
	mutex     sync.Mutex
```

File: pkg/controller/job-controller.go [hunk 8]
```diff
@@ -203,16 +285,428 @@ func (c *JobController) SyncHandler(key string) (Result, error) {
...
+
+func (intervalJob *MinIOIntervalJob) createCommandJob(ctx context.Context, k8sClient client.Client) error {
+	for _, command := range intervalJob.Command {
+		if len(command.DepnedOn) == 0 {
+			err := command.createJob(ctx, k8sClient, intervalJob.JobCR)
+			if err != nil {
+				return err
+			}
+		} else {
+			dependAllSucess := true
```
Comment (shtripat) [inline pkg/controller/job-controller.go:554]:
```suggestion
			allDepsSuccess := true
```

File: pkg/controller/job-controller.go [hunk 9]
```diff
@@ -203,16 +285,428 @@ func (c *JobController) SyncHandler(key string) (Result, error) {
...
+			for _, dep := range command.DepnedOn {
+				status, found := intervalJob.CommandMap[dep]
+				if !found {
+					return fmt.Errorf("depended job %s not found", dep)
```
Comment (shtripat) [inline pkg/controller/job-controller.go:558]:
```suggestion
					return fmt.Errorf("dependent job %s not found", dep)
```

File: pkg/controller/job-controller.go [hunk 10]
```diff
@@ -203,16 +285,428 @@ func (c *JobController) SyncHandler(key string) (Result, error) {
...
+		if !found {
+			return intervalJob, fmt.Errorf("operation %s is not supported", val.Operation)
+		}
+		var command string
+		argsTpl, found := jobOperation[mcCommand]
+		if !found {
+			return intervalJob, fmt.Errorf("operation %s is not supported", mcCommand)
+		}
+		// [ALIAS]
+		command = strings.ReplaceAll(argsTpl, "[ALIAS]", "myminio")
```
Comment (shtripat) [inline pkg/controller/job-controller.go:615]: Should we use template package here?

File: pkg/controller/job-controller.go [hunk 11]
```diff
@@ -203,16 +285,428 @@ func (c *JobController) SyncHandler(key string) (Result, error) {
...
+		flags := []string{}
+		for flagsName, flagsVal := range val.Args {
```
Comment (shtripat) [inline pkg/controller/job-controller.go:617]: flagsName and flagVal can be singular like `flagName` and `flagVal`

File: pkg/controller/job-controller.go [hunk 12]
```diff
@@ -203,16 +285,428 @@ func (c *JobController) SyncHandler(key string) (Result, error) {
...
+			Files:       []MinIOIntervalJobCommandFile{},
+		}
+		// some commands need to have a empty name
+		if jobCommand.JobName == "" {
+			jobCommand.JobName = fmt.Sprintf("command-%d", index)
+		}
+		var matchString, name, ext string
+		found = true
+		for found {
+			matchString, name, ext, found = getFileNameAndExt(command)
```
Comment (shtripat) [inline pkg/controller/job-controller.go:660]:
```suggestion
			matchString, name, ext, cmdFound = getFileNameAndExt(command)
```

File: pkg/controller/job-controller.go [hunk 13]
```diff
@@ -203,16 +285,428 @@ func (c *JobController) SyncHandler(key string) (Result, error) {
...
+		}
+		intervalJob.Command = append(intervalJob.Command, &jobCommand)
+		intervalJob.CommandMap[jobCommand.JobName] = &jobCommand
+	}
+	// check all dependon
+	for _, command := range intervalJob.Command {
+		for _, dep := range command.DepnedOn {
+			_, found := intervalJob.CommandMap[dep]
+			if !found {
+				return intervalJob, fmt.Errorf("depended job %s not found", dep)
```
Comment (shtripat) [inline pkg/controller/job-controller.go:684]:
```suggestion
				return intervalJob, fmt.Errorf("dependent job %s not found", dep)
```

File: pkg/controller/job-controller.go [hunk 14]
```diff
@@ -203,16 +285,428 @@ func (c *JobController) SyncHandler(key string) (Result, error) {
...
+
+func getFileNameAndExt(commond string) (matchString, name, ext string, found bool) {
+	match := fileRegexp.MatchString(commond)
+	if !match {
+		return "", "", "", false
+	}
+	matchStrings := fileRegexp.FindAllStringSubmatch(commond, -1)
+	for _, val := range matchStrings {
+		if len(val) == 2 {
+			// val[1] must be `[file(policy,json)]`
```
Comment (shtripat) [inline pkg/controller/job-controller.go:700]: `[file(policy.json)]` right?

File: pkg/controller/job-controller.go (hunk 10 above)
Comment (jiuker) [reply pkg/controller/job-controller.go:614]: I have another idea for this. But you can test this first.

File: pkg/controller/job-controller.go (hunk 2 above)
Comment (shtripat) [reply pkg/controller/job-controller.go:243]: comment says `do nothing` but below lines delete, so we can remove the comment

File: pkg/utils/miniojob/minioJob.go [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+	FileContext string
+}
+
+// IsFile - if it is a file
+func (arg Arg) IsFile() bool {
+	return arg.FileName != ""
+}
+
+// FiledsFunc - alias function
+type FiledsFunc func(args map[string]string) (Arg, error)
```
Comment (shtripat) [inline pkg/utils/miniojob/minioJob.go:39]:
```suggestion
type FieldsFunc func(args map[string]string) (Arg, error)
```

File: pkg/utils/miniojob/minioJob.go [hunk 16]
```diff
@@ -0,0 +1,170 @@
...
+
+// Key - key=value
+func Key(key string) FiledsFunc {
+	return KeyForamt(key, "$0")
+}
+
+// FLAGS - --key=value|value1,value2,value3
+func FLAGS(igrnoreKeys ...string) FiledsFunc {
```
Comment (shtripat) [inline pkg/utils/miniojob/minioJob.go:47]:
```suggestion
func FLAGS(ignoreKeys ...string) FieldsFunc {
```

**Review by shtripat — DISMISSED** [review]
Verified the changes and works as expected.
Few minor ones.

**Review by cniackz — DISMISSED** [review]

File: pkg/utils/miniojob/minioJob.go [hunk 17]
```diff
@@ -0,0 +1,170 @@
...
+			if out, err = func1(args); err == nil {
+				return out, nil
+			}
+		}
+		return out, fmt.Errorf("not found")
+	}
+}
+
+// NoSpace - no space for the command
+func NoSpace(funcs ...FiledsFunc) FiledsFunc {
```
[the remark this answers is not in the document]
Comment (jiuker) [reply pkg/utils/miniojob/minioJob.go]: I think that's ok. For Every fileds will generate like `a b c d`. We wrap them can get `abcd`. For `mc mb alias/bucket`. Any name for this?

**Review by shtripat — APPROVED** [review]

**Review by cniackz — APPROVED** [review]
LGTM

**Pull request merged**